File: cocos-creator-project/assets/scripts/game/ThreatBoard.ts
import { TrackClassification, GameConstants, getTrackIdString, getClassificationString } from './GameTypes';
import { TrackManager } from './TrackManager';

// ============================================================================
// Threat Assessment Entry — one row on the threat board (Scope 2)
// ============================================================================

export interface ThreatEntry {
    trackId: number;
    trackIdString: string;
    classification: string;
    range: number;
    azimuth: number;
    altitude: number;
    speed: number;
    heading: number;
    inbound: boolean;
    closingRate: number;
    threatScore: number;
    timeToTerritory: number;
}

// ============================================================================
// ThreatBoard — the second scope heads-up display
// ============================================================================

export class ThreatBoard {
    static readonly MAX_DISPLAYED_THREATS = 5;

    private topThreats: ThreatEntry[] = [];

    update(trackManager: TrackManager): void {
        const allTracks = trackManager.getAllTracks();
        const candidates: ThreatEntry[] = [];

        for (const track of allTracks) {
            if (track.classification === TrackClassification.FRIENDLY) continue;
            if (!track.isAlive) continue;

            const inbound = ThreatBoard.isInbound(track.azimuth, track.heading);
            const closingRate = ThreatBoard.calculateClosingRate(
                track.range, track.azimuth, track.speed, track.heading);
            const threatScore = ThreatBoard.calculateThreatScore(track);
            const timeToTerritory = ThreatBoard.estimateTimeToTerritory(track.range, closingRate);

            candidates.push({
                trackId: track.trackId,
                trackIdString: getTrackIdString(track.trackId),
                classification: getClassificationString(track.classification),
                range: track.range,
                azimuth: track.azimuth,
                altitude: track.altitude,
                speed: track.speed,
                heading: track.heading,
                inbound,
                closingRate,
                threatScore,
                timeToTerritory,
            });
        }

        candidates.sort((a, b) => b.threatScore - a.threatScore);
        this.topThreats = candidates.slice(0, ThreatBoard.MAX_DISPLAYED_THREATS);
    }

    getTopThreats(): ThreatEntry[] { return this.topThreats; }
    getThreatCount(): number { return this.topThreats.length; }

    getThreat(rank: number): ThreatEntry | null {
        return (rank >= 0 && rank < this.topThreats.length) ? this.topThreats[rank] : null;
    }

    isOnBoard(trackId: number): boolean {
        return this.topThreats.some((t) => t.trackId === trackId);
    }

    getThreatByTrackId(trackId: number): ThreatEntry | null {
        return this.topThreats.find((t) => t.trackId === trackId) || null;
    }

    formatBoard(): string {
        let board = '=== THREAT BOARD (SCOPE 2) ===\n';
        board += `  TOP ${ThreatBoard.MAX_DISPLAYED_THREATS} THREATS\n`;
        board += '------------------------------\n';

        if (this.topThreats.length === 0) {
            board += '  NO THREATS DETECTED\n';
            return board;
        }

        for (let i = 0; i < this.topThreats.length; i++) {
            const t = this.topThreats[i];
            const eta = t.timeToTerritory < 9000 ? `${Math.floor(t.timeToTerritory)}s` : 'N/A';
            board += `#${i + 1} ${t.trackIdString} ${t.classification} ` +
                     `RNG:${Math.floor(t.range)}km AZ:${String(Math.floor(t.azimuth)).padStart(3, '0')} ` +
                     `SPD:${Math.floor(t.speed)}kts ${t.inbound ? 'INBOUND' : 'OUTBND '} ETA:${eta}\n`;
        }

        return board;
    }

    private static calculateThreatScore(track: { range: number; speed: number; azimuth: number; heading: number; classification: TrackClassification; altitude: number }): number {
        let score = 0;

        const proximityScore = ((GameConstants.RADAR_MAX_RANGE_KM - track.range) /
            GameConstants.RADAR_MAX_RANGE_KM) * 100.0;
        score += proximityScore * 3.0;

        const speedScore = (track.speed / 1000.0) * 50.0;
        score += speedScore;

        if (ThreatBoard.isInbound(track.azimuth, track.heading)) {
            score += 150.0;
        }

        if (track.classification === TrackClassification.HOSTILE) {
            score += 100.0;
        } else if (track.classification === TrackClassification.UNKNOWN) {
            score += 50.0;
        }

        if (track.altitude < 5000) {
            score += 30.0;
        }

        if (track.range <= GameConstants.TERRITORY_RADIUS_KM) {
            score += 200.0;
        }

        return score;
    }

    static isInbound(azimuth: number, heading: number): boolean {
        const inboundHeading = (azimuth + 180.0) % 360.0;
        let diff = ((Math.abs(heading - inboundHeading) + 360.0) % 360.0);
        if (diff > 180.0) diff = 360.0 - diff;
        return diff <= 60.0;
    }

    private static calculateClosingRate(range: number, azimuth: number, speed: number, heading: number): number {
        const speedKmPerSec = (speed * GameConstants.NM_TO_KM) / 3600.0;
        const inboundHeading = (azimuth + 180.0) % 360.0;
        const diff = ((heading - inboundHeading) * Math.PI) / 180.0;
        return speedKmPerSec * Math.cos(diff);
    }

    private static estimateTimeToTerritory(range: number, closingRate: number): number {
        if (closingRate <= 0.001) return 9999.0;
        const distToTerritory = range - GameConstants.TERRITORY_RADIUS_KM;
        if (distToTerritory <= 0) return 0;
        return distToTerritory / closingRate;
    }
}

File: src/ThreatBoard.cpp
#include "ThreatBoard.h"
#include <algorithm>
#include <cmath>
#include <cstdio>

ThreatBoard::ThreatBoard() {}
ThreatBoard::~ThreatBoard() {}

void ThreatBoard::update(const TrackManager& trackManager)
{
    // Get all current tracks
    auto allTracks = const_cast<TrackManager&>(trackManager).getAllTracks();

    // Build threat entries for non-friendly tracks
    std::vector<ThreatEntry> candidates;

    for (const auto& track : allTracks) {
        // Skip friendly and dead tracks
        if (track.classification == TrackClassification::FRIENDLY) continue;
        if (!track.isAlive) continue;

        ThreatEntry entry;
        entry.trackId = track.trackId;
        entry.trackIdString = track.getTrackIdString();
        entry.classification = track.getClassificationString();
        entry.range = track.range;
        entry.azimuth = track.azimuth;
        entry.altitude = track.altitude;
        entry.speed = track.speed;
        entry.heading = track.heading;
        entry.inbound = isInbound(track.azimuth, track.heading);
        entry.closingRate = calculateClosingRate(track.range, track.azimuth,
                                                  track.speed, track.heading);
        entry.threatScore = calculateThreatScore(track);
        entry.timeToTerritory = estimateTimeToTerritory(track.range, entry.closingRate);

        candidates.push_back(entry);
    }

    // Sort by threat score (highest first)
    std::sort(candidates.begin(), candidates.end(),
        [](const ThreatEntry& a, const ThreatEntry& b) {
            return a.threatScore > b.threatScore;
        });

    // Keep top N
    topThreats_.clear();
    int count = std::min((int)candidates.size(), MAX_DISPLAYED_THREATS);
    for (int i = 0; i < count; i++) {
        topThreats_.push_back(candidates[i]);
    }
}

const ThreatEntry* ThreatBoard::getThreat(int rank) const
{
    if (rank < 0 || rank >= (int)topThreats_.size()) return nullptr;
    return &topThreats_[rank];
}

bool ThreatBoard::isOnBoard(int trackId) const
{
    for (const auto& t : topThreats_) {
        if (t.trackId == trackId) return true;
    }
    return false;
}

const ThreatEntry* ThreatBoard::getThreatByTrackId(int trackId) const
{
    for (const auto& t : topThreats_) {
        if (t.trackId == trackId) return &t;
    }
    return nullptr;
}

float ThreatBoard::calculateThreatScore(const TrackData& track)
{
    float score = 0.0f;

    // Proximity — closer targets are higher threat
    // Invert range: max radar range minus current range
    float proximityScore = (GameConstants::RADAR_MAX_RANGE_KM - track.range)
                           / GameConstants::RADAR_MAX_RANGE_KM * 100.0f;
    score += proximityScore * 3.0f;  // Proximity is weighted heavily

    // Speed — faster targets need faster response
    float speedScore = track.speed / 1000.0f * 50.0f;  // Normalized against ~1000 kts
    score += speedScore;

    // Inbound bonus — heading toward radar center is far more threatening
    if (isInbound(track.azimuth, track.heading)) {
        score += 150.0f;
    }

    // Classification bonus
    if (track.classification == TrackClassification::HOSTILE) {
        score += 100.0f;  // Confirmed hostile gets priority
    } else if (track.classification == TrackClassification::UNKNOWN) {
        score += 50.0f;   // Unknown is still concerning
    }

    // Altitude factor — low altitude penetrators are harder to detect and engage
    if (track.altitude < 5000.0f) {
        score += 30.0f;  // Low-altitude bonus
    }

    // Inside territory radius — critical threat
    if (track.range <= GameConstants::TERRITORY_RADIUS_KM) {
        score += 200.0f;
    }

    return score;
}

bool ThreatBoard::isInbound(float azimuth, float heading)
{
    // An aircraft is inbound if its heading is roughly opposite to its azimuth
    // from radar center. Azimuth is where the aircraft IS (from center),
    // heading is where it's GOING.
    //
    // If aircraft is at azimuth 90 (due east), an inbound heading would be
    // roughly 270 (heading west, toward center).
    float inboundHeading = std::fmod(azimuth + 180.0f, 360.0f);
    float diff = std::fmod(std::abs(heading - inboundHeading) + 360.0f, 360.0f);
    if (diff > 180.0f) diff = 360.0f - diff;

    // Within 60 degrees of direct inbound heading
    return diff <= 60.0f;
}

float ThreatBoard::calculateClosingRate(float range, float azimuth,
                                          float speed, float heading)
{
    // Convert speed from knots to km/s
    float speedKmPerSec = speed * GameConstants::NM_TO_KM / 3600.0f;

    // Calculate radial component of velocity (positive = closing)
    float inboundHeading = std::fmod(azimuth + 180.0f, 360.0f);
    float diff = (heading - inboundHeading) * M_PI / 180.0f;
    float radialComponent = speedKmPerSec * std::cos(diff);

    return radialComponent;
}

float ThreatBoard::estimateTimeToTerritory(float range, float closingRate)
{
    if (closingRate <= 0.001f) return 9999.0f;  // Not closing

    float distToTerritory = range - GameConstants::TERRITORY_RADIUS_KM;
    if (distToTerritory <= 0.0f) return 0.0f;   // Already inside

    return distToTerritory / closingRate;
}

std::string ThreatEntry::formatLine() const
{
    char buf[256];
    snprintf(buf, sizeof(buf),
        "%s %s RNG:%.0fkm AZ:%03d SPD:%.0fkts %s ETA:%s",
        trackIdString.c_str(),
        classification.c_str(),
        range, (int)azimuth, speed,
        inbound ? "INBOUND" : "OUTBND ",
        timeToTerritory < 9000.0f ?
            (std::to_string((int)timeToTerritory) + "s").c_str() : "N/A");
    return std::string(buf);
}

std::string ThreatEntry::formatCard() const
{
    char buf[512];
    snprintf(buf, sizeof(buf),
        "  TRACK: %s  [%s]\n"
        "  RNG:   %.1f km (%.1f NM)\n"
        "  AZM:   %03d deg\n"
        "  ALT:   %.0f ft\n"
        "  SPD:   %.0f kts\n"
        "  HDG:   %03d deg\n"
        "  DIR:   %s\n"
        "  CLOSE: %.2f km/s\n"
        "  ETA:   %s\n"
        "  SCORE: %.0f",
        trackIdString.c_str(), classification.c_str(),
        range, range * GameConstants::KM_TO_NM,
        (int)azimuth,
        altitude,
        speed,
        (int)heading,
        inbound ? "INBOUND" : "OUTBOUND",
        closingRate,
        timeToTerritory < 9000.0f ?
            (std::to_string((int)timeToTerritory) + "s").c_str() : "N/A",
        threatScore);
    return std::string(buf);
}

std::string ThreatBoard::formatBoard() const
{
    std::string board;
    board += "=== THREAT BOARD (SCOPE 2) ===\n";
    board += "  TOP " + std::to_string(MAX_DISPLAYED_THREATS) + " THREATS\n";
    board += "------------------------------\n";

    if (topThreats_.empty()) {
        board += "  NO THREATS DETECTED\n";
        return board;
    }

    for (int i = 0; i < (int)topThreats_.size(); i++) {
        char header[32];
        snprintf(header, sizeof(header), "#%d ", i + 1);
        board += header;
        board += topThreats_[i].formatLine() + "\n";
    }

    return board;
}
